### PR TITLE
Add unregister device operation

### PR DESCRIPTION
**Pull request by esdrasjnr**
What does this implement/fix? Explain your changes
---------------------------------------------------
This patch implements the operation of unregister devices on babeltower service and repass this same command to connector.

Where has this been tested?
---------------------------
**Operating System/Platform:** Ubuntu 18.04.4

**Go Version:** 1.13.6

### CODE REVIEW

File: pkg/thing/interactors/unregister_thing.go [hunk 1]
```diff
@@ -0,0 +1,39 @@
+package interactors
+
+import (
+	"errors"
+)
+
+func (i *ThingInteractor) Unregister(authorization, id string) error {
```
Comment (joaoaneto) [inline pkg/thing/interactors/unregister_thing.go]: Codacy Issue found: [exported method ThingInteractor.Unregister should have comment or be unexported](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5019174)

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- pkg/thing/interactors/auth_thing_test.go  5
- pkg/thing/handler/amqp/msg_handler.go  4
- pkg/thing/interactors/unregister_thing.go  7
- pkg/thing/entities/err_thing_forbidden.go  1
- pkg/thing/interactors/auth_thing.go  6
- pkg/thing/entities/err_thing_not_found.go  1
- pkg/thing/interactors/unregister_thing_test.go  3
         

Clones added
============
- pkg/thing/interactors/auth_thing_test.go  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5019174&bid=16546397)

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- pkg/thing/interactors/auth_thing_test.go  5
- pkg/thing/handler/amqp/msg_handler.go  4
- pkg/thing/entities/err_thing_forbidden.go  1
- pkg/thing/interactors/unregister_thing.go  7
- pkg/thing/entities/err_thing_not_found.go  1
- pkg/thing/interactors/unregister_thing_test.go  3
- pkg/thing/interactors/auth_thing.go  6
         

Clones added
============
- pkg/thing/interactors/auth_thing_test.go  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5019174&bid=16546397)

File: pkg/thing/interactors/unregister_thing.go (hunk 1 above)
Comment (joaoaneto) [inline pkg/thing/interactors/unregister_thing.go]: Codacy Issue found: [exported method ThingInteractor.Unregister should have comment or be unexported](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5019174)

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/thing/interactors/unregister_thing.go  7
- pkg/thing/interactors/unregister_thing_test.go  3
- pkg/thing/entities/err_thing_not_found.go  1
- pkg/thing/interactors/auth_thing.go  6
- pkg/thing/entities/err_thing_forbidden.go  1
- pkg/thing/handler/amqp/msg_handler.go  4
- pkg/thing/interactors/auth_thing_test.go  5
         

Clones added
============
- pkg/thing/interactors/auth_thing_test.go  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5019174&bid=16546397)

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/thing/interactors/auth_thing.go  6
- pkg/thing/interactors/unregister_thing_test.go  3
- pkg/thing/interactors/unregister_thing.go  7
- pkg/thing/entities/err_thing_forbidden.go  1
- pkg/thing/handler/amqp/msg_handler.go  4
- pkg/thing/entities/err_thing_not_found.go  1
- pkg/thing/interactors/auth_thing_test.go  5
         

Clones added
============
- pkg/thing/interactors/auth_thing_test.go  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5019174&bid=16546397)

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/thing/interactors/auth_thing.go  6
- pkg/thing/interactors/unregister_thing_test.go  3
- pkg/thing/entities/err_thing_not_found.go  1
- pkg/thing/entities/err_thing_forbidden.go  1
- pkg/thing/interactors/unregister_thing.go  7
- pkg/thing/interactors/auth_thing_test.go  5
- pkg/thing/handler/amqp/msg_handler.go  4
         

Clones added
============
- pkg/thing/interactors/auth_thing_test.go  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5019174&bid=16546397)

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/thing/entities/err_thing_not_found.go  1
- pkg/thing/interactors/unregister_thing.go  7
- pkg/thing/interactors/unregister_thing_test.go  3
- pkg/thing/entities/err_thing_forbidden.go  1
- pkg/thing/interactors/auth_thing.go  6
- pkg/thing/interactors/auth_thing_test.go  5
- pkg/thing/handler/amqp/msg_handler.go  4
         

Clones added
============
- pkg/thing/interactors/auth_thing_test.go  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5019174&bid=16546397)

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/thing/interactors/unregister_thing_test.go  3
- pkg/thing/interactors/auth_thing_test.go  5
- pkg/thing/handler/amqp/msg_handler.go  4
- pkg/thing/entities/err_thing_forbidden.go  1
- pkg/thing/interactors/unregister_thing.go  7
- pkg/thing/interactors/auth_thing.go  6
- pkg/thing/entities/err_thing_not_found.go  1
         

Clones added
============
- pkg/thing/interactors/auth_thing_test.go  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5019174&bid=16546397)

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/thing/interactors/auth_thing.go  6
- pkg/thing/interactors/auth_thing_test.go  5
- pkg/thing/interactors/unregister_thing.go  7
- pkg/thing/handler/amqp/msg_handler.go  4
- pkg/thing/interactors/unregister_thing_test.go  3
- pkg/thing/entities/err_thing_not_found.go  1
- pkg/thing/entities/err_thing_forbidden.go  1
         

Clones added
============
- pkg/thing/interactors/auth_thing_test.go  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5019174&bid=16546397)

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/thing/handler/amqp/msg_handler.go  2
- pkg/thing/interactors/unregister_thing_test.go  3
- pkg/thing/interactors/unregister_thing.go  7
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5019174&bid=16546397)

File: pkg/thing/delivery/http/thing_proxy.go [hunk 2]
```diff
@@ -266,3 +267,29 @@ func (p proxy) GetThing(authorization, ID string) (*entities.Thing, error) {
...
+		nil,
+		nil,
+	}
+
+	resp, err := p.sendRequest(requestInfo)
+	if err != nil {
+		return err
+	}
+
+	defer resp.Body.Close()
```
Comment (lcbm) [inline pkg/thing/delivery/http/thing_proxy.go:293]: As mentioned [here](https://github.com/CESARBR/knot-babeltower/pull/42#discussion_r386715896), in order to maintain code standardization, there should be no blank lines between the 'defer' statement and the function it refers to. :smile:

File: pkg/thing/delivery/amqp/client.go [hunk 3]
```diff
@@ -26,6 +27,7 @@ type msgClientPublisher struct {
 // ClientPublisher is the interface with methods that the publisher should have
 type ClientPublisher interface {
 	SendRegisterDevice(network.RegisterResponseMsg) error
+	SendUnregisterDevice(thingID string, errMsg *string) error
```
Comment (netoax) [inline pkg/thing/delivery/amqp/client.go]: I would prefer to rename this method to `SendUnregisteredDevice` because we are sending an `unregistered` event. `SendRegisterDevice` should be renamed too, IMO.

File: pkg/thing/delivery/amqp/client.go [hunk 4]
```diff
@@ -50,6 +52,20 @@ func (mp *msgClientPublisher) SendRegisterDevice(msg network.RegisterResponseMsg
 	return mp.amqp.PublishPersistentMessage(exchangeFogOut, registerOutKey, jsonMsg)
 }
 
+// SendUnregisterDevice sends an unregistered message
+func (mp *msgClientPublisher) SendUnregisterDevice(thingID string, errMsg *string) error {
+	mp.logger.Debug("Sending unregister message")
+
```
Comment (netoax) [inline pkg/thing/delivery/amqp/client.go]: I think it's better to remove this blank line.

File: pkg/thing/delivery/amqp/connector.go [hunk 5]
```diff
@@ -44,6 +46,20 @@ func (mp *msgConnectorPublisher) SendRegisterDevice(id string, name string) erro
 	return mp.amqp.PublishPersistentMessage(exchangeConnIn, registerInKey, bytes)
 }
 
+// SendUnregisterDevice sends an unregister message
+func (mp *msgConnectorPublisher) SendUnregisterDevice(id string) error {
+	mp.logger.Debug("Sending unregister message")
+
```
Comment (netoax) [inline pkg/thing/delivery/amqp/connector.go]: Same about the blank line.

File: pkg/thing/handler/amqp/msg_handler.go [hunk 6]
```diff
@@ -79,6 +79,16 @@ func (mc *MsgHandler) handleRegisterMsg(body []byte, authorizationHeader string)
 	return mc.thingInteractor.Register(authorizationHeader, msgParsed.ID, msgParsed.Name)
 }
 
+func (mc *MsgHandler) handleUnregisterMsg(body []byte, authorizationHeader string) error {
+	msgParsed := network.UnregisterRequestMsg{}
```
Comment (netoax) [inline pkg/thing/handler/amqp/msg_handler.go]: As the function is small, I think this variable could be renamed to `msg`.

File: pkg/thing/interactors/unregister_thing.go [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+		sendErr := i.clientPublisher.SendUnregisterDevice(id, &msg)
+		if sendErr != nil {
+			return sendErr
+		}
+		return err
+	}
+
+	sendErr := i.connectorPublisher.SendUnregisterDevice(id)
+	if sendErr != nil {
+		i.logger.Error(sendErr)
```
Comment (netoax) [inline pkg/thing/interactors/unregister_thing.go:33]: Shouldn't the error be returned?

**Review by netoax — CHANGES_REQUESTED** [review]
Thanks for your PR, it's very good. Just some adjustments.

Also:
- Commit title (940381135d53285ff373a7b29b5c018e9661b50a) could be changed to **comments: Fix syntax errors**. It's clear and avoids the need of a commit message.
- Commit title (c4702cacbd84b6f39ebdacea981baacf0c0555fc and 81d826760e416f1ce64c7c8607367ae4a3a3e1e6): prefix client and connect are enough, without **publisher**.

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/thing/handler/amqp/msg_handler.go  2
- pkg/thing/interactors/unregister_thing_test.go  3
- pkg/thing/interactors/unregister_thing.go  7
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5019174&bid=16546397)

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
           

Complexity increasing per file
==============================
- pkg/thing/interactors/unregister_thing.go  7
- pkg/thing/handler/amqp/msg_handler.go  2
- pkg/thing/interactors/unregister_thing_test.go  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5019174&bid=16546397)

File: pkg/thing/delivery/amqp/client.go (hunk 3 above)
Comment (esdrasjnr) [reply pkg/thing/delivery/amqp/client.go]: You are right @joaoaneto! Fixing the `unregister` in this commit, in another I've fixed for `register` too. Done!

File: pkg/thing/delivery/amqp/client.go (hunk 4 above)
Comment (esdrasjnr) [reply pkg/thing/delivery/amqp/client.go]: @joaoaneto Done!

File: pkg/thing/delivery/amqp/connector.go (hunk 5 above)
Comment (esdrasjnr) [reply pkg/thing/delivery/amqp/connector.go]: @joaoaneto Done!

File: pkg/thing/handler/amqp/msg_handler.go (hunk 6 above)
Comment (esdrasjnr) [reply pkg/thing/handler/amqp/msg_handler.go]: @joaoaneto Agree with you, done!

File: pkg/thing/interactors/unregister_thing.go (hunk 7 above)
Comment (esdrasjnr) [reply pkg/thing/interactors/unregister_thing.go:33]: @joaoaneto In this case, return the function at this point will avoid sending a client request response. However the error at sending a connector request should be treated right. By now I add a `TODO` at this code point to analyse carefully the possible errors and what should be done.

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
           

Complexity increasing per file
==============================
- pkg/thing/interactors/unregister_thing_test.go  3
- pkg/thing/handler/amqp/msg_handler.go  2
- pkg/thing/interactors/unregister_thing.go  7
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5019174&bid=16546397)

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
           

Complexity increasing per file
==============================
- pkg/thing/interactors/unregister_thing_test.go  3
- pkg/thing/interactors/unregister_thing.go  7
- pkg/thing/handler/amqp/msg_handler.go  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5019174&bid=16546397)

File: pkg/thing/handler/amqp/msg_handler.go [hunk 8]
```diff
@@ -79,6 +79,17 @@ func (mc *MsgHandler) handleRegisterMsg(body []byte, authorizationHeader string)
 	return mc.thingInteractor.Register(authorizationHeader, msgParsed.ID, msgParsed.Name)
 }
 
+func (mc *MsgHandler) handleUnregisterMsg(body []byte, authorizationHeader string) error {
+	msg := network.UnregisterRequestMsg{}
```
Comment (lcbm) [inline pkg/thing/handler/amqp/msg_handler.go:83]: Considering `msg` was renamed from `msgParsed` (in response to [this](https://github.com/CESARBR/knot-babeltower/pull/41#discussion_r388302550) comment), we should add a commit to rename `msgParsed` at **line 73**. :smile:

File: pkg/thing/interactors/unregister_thing.go [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+	}
+
+	err := i.thingProxy.RemoveThing(authorization, id)
+	if err != nil {
+		msg := err.Error()
+		sendErr := i.clientPublisher.SendUnregisteredDevice(id, &msg)
+		if sendErr != nil {
+			return sendErr
+		}
+		return err
```
Comment (lcbm) [inline pkg/thing/interactors/unregister_thing.go:27]: Don't forget to add `i.logger.Error(sendErr)` before returning these errors :stuck_out_tongue:

File: pkg/thing/interactors/unregister_thing.go [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+
+	sendErr := i.connectorPublisher.SendUnregisterDevice(id)
+	if sendErr != nil {
+		// TODO handle error at sending unregister message to connector
+		i.logger.Error(sendErr)
+	}
+
+	sendErr = i.clientPublisher.SendUnregisteredDevice(id, nil)
+	if sendErr != nil {
+		return sendErr
```
Comment (lcbm) [inline pkg/thing/interactors/unregister_thing.go:38]: Same as [here]()! Don't forget to add i.logger.Error(sendErr) before returning the error :smile:

File: pkg/network/msg.go [hunk 11]
```diff
@@ -8,6 +8,11 @@ type RegisterRequestMsg struct {
 	Name string `json:"name"`
 }
 
+// UnregisterRequestMsg is received to unregister a device
```
Comment (lcbm) [inline pkg/network/msg.go]: Do you think we change the wording a little so it matches the pattern used to comment on other `types`? :grimacing: 

Here's an example :point_down:  
> `// UnregisterRequestMsg represents the incoming unregister device request message.`

File: pkg/network/msg.go [hunk 12]
```diff
@@ -21,6 +26,12 @@ type RegisterResponseMsg struct {
 	Error *string `json:"error"`
 }
 
+// UnregisterResponseMsg is sent when receive a unregister request
```
Comment (lcbm) [inline pkg/network/msg.go]: Just as `UnregisterRequestMsg`, what do you think about changing the wording a little so it matches the pattern used to comment on other `types`? :thinking: 

Here's another example :point_down:  
> `// UnregisterResponseMsg represents the outgoing unregister device response message.`

File: pkg/thing/delivery/amqp/client.go [hunk 13]
```diff
@@ -37,10 +39,9 @@ func NewMsgClientPublisher(logger logging.Logger, amqp *network.Amqp) ClientPubl
 	return &msgClientPublisher{logger, amqp}
 }
 
-// SendRegisterDevice sends a registered message
-func (mp *msgClientPublisher) SendRegisterDevice(msg network.RegisterResponseMsg) error {
-	mp.logger.Debug("Sending register message")
-
+// SendRegisteredDevice sends a registered message
```
Comment (lcbm) [inline pkg/thing/delivery/amqp/client.go]: Do you think we be more descriptive here? :thinking: 
What do you think about changing to something along these lines? :point_down: 
> `// SendRegisteredDevice publishes the registered device's credentials to the device registration queue.`

File: pkg/thing/delivery/amqp/client.go [hunk 14]
```diff
@@ -50,6 +51,19 @@ func (mp *msgClientPublisher) SendRegisterDevice(msg network.RegisterResponseMsg
 	return mp.amqp.PublishPersistentMessage(exchangeFogOut, registerOutKey, jsonMsg)
 }
 
+// SendUnregisterDevice sends an unregistered message
```
Comment (lcbm) [inline pkg/thing/delivery/amqp/client.go]: Just like the prior review, do you think we would benefit from being a little more descriptive? :eyes:
 
Something along these lines would, maybe? :point_down: 
> `// SendUnregisterDevice publishes the unregistered device's id and error message to the device unregistered queue.`

**Review by lcbm — CHANGES_REQUESTED** [review]
Awesome PR :heart_eyes: 

Most comments I left were suggestions related to code commenting and code standardization. However, there is some logging missing :eyes: 

Let me know when you take a look at it :smile:

File: pkg/thing/delivery/http/thing_proxy.go (hunk 2 above)
Comment (esdrasjnr) [reply pkg/thing/delivery/http/thing_proxy.go:293]: @lcbm You're right! I've added this fix in a patch above this PR. Check this [comment](https://github.com/CESARBR/knot-babeltower/pull/42#discussion_r386715896) :smiley:

File: pkg/thing/handler/amqp/msg_handler.go (hunk 8 above)
Comment (esdrasjnr) [reply pkg/thing/handler/amqp/msg_handler.go:83]: @lcbm I've done this in the latest commit of this PR! Check it [here](https://github.com/CESARBR/knot-babeltower/pull/41/commits/846770a64843e1c543740ba9b295da361bce6daf) :smiley:

File: pkg/thing/interactors/unregister_thing.go (hunk 9 above)
Comment (esdrasjnr) [reply pkg/thing/interactors/unregister_thing.go:27]: @lcbm I didn't add a log before any of the error returns. This is because the method whose calls this interactor (in `msg_hander.go`) logs every returned error, so I thought that log the error in interactor itself would be ambiguous.
However, in that case I think that `msg` (line 21) should be notified by logger before returning `sendErr` (line 24)

File: pkg/thing/interactors/unregister_thing.go (hunk 10 above)
Comment (esdrasjnr) [reply pkg/thing/interactors/unregister_thing.go:38]: Same as I commented [here](https://github.com/CESARBR/knot-babeltower/pull/41#discussion_r389854783) :smiley:

File: pkg/network/msg.go (hunk 11 above)
Comment (esdrasjnr) [reply pkg/network/msg.go]: @lcbm Agree with you, done!

File: pkg/network/msg.go (hunk 12 above)
Comment (esdrasjnr) [reply pkg/network/msg.go]: @lcbm Done!

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
           

Complexity increasing per file
==============================
- pkg/thing/handler/amqp/msg_handler.go  2
- pkg/thing/interactors/unregister_thing_test.go  3
- pkg/thing/interactors/unregister_thing.go  7
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5019174&bid=16546397)

File: pkg/thing/delivery/amqp/client.go (hunk 13 above)
Comment (esdrasjnr) [reply pkg/thing/delivery/amqp/client.go]: @lcbm Done!

File: pkg/thing/delivery/amqp/client.go (hunk 14 above)
Comment (esdrasjnr) [reply pkg/thing/delivery/amqp/client.go]: @lcbm Done!

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
           

Complexity increasing per file
==============================
- pkg/thing/interactors/unregister_thing.go  7
- pkg/thing/handler/amqp/msg_handler.go  2
- pkg/thing/interactors/unregister_thing_test.go  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5019174&bid=16546397)

File: pkg/network/msg.go [hunk 15]
```diff
@@ -2,12 +2,17 @@ package network
 
 import "github.com/CESARBR/knot-babeltower/pkg/thing/entities"
 
-// RegisterRequestMsg is received to register a device
+// RegisterRequestMsg represents the incoming register device request message
```
Comment (netoax) [inline pkg/network/msg.go:5]: This change should be done in another commit.

File: pkg/thing/delivery/http/thing_proxy.go [hunk 16]
```diff
@@ -267,3 +268,29 @@ func (p proxy) GetThing(authorization, ID string) (*entities.Thing, error) {
 
 	return nil, &entities.ErrThingNotFound{ID: ID}
 }
+
+// RemoveThing remove the indicated thing from service
+func (p proxy) RemoveThing(authorization, ID string) error {
```
Comment (netoax) [inline pkg/thing/delivery/http/thing_proxy.go]: IMO, you can rename this method to `Remove()` since the operation is already on things context.
Please, create another commit to rename the method `GetThing()` too.

File: pkg/thing/delivery/http/thing_proxy.go [hunk 17]
```diff
@@ -267,3 +268,29 @@ func (p proxy) GetThing(authorization, ID string) (*entities.Thing, error) {
 
 	return nil, &entities.ErrThingNotFound{ID: ID}
 }
+
+// RemoveThing remove the indicated thing from service
```
Comment (netoax) [inline pkg/thing/delivery/http/thing_proxy.go]: **removes** the indicated thing from **the thing's service**

File: pkg/network/msg.go [hunk 18]
```diff
@@ -19,13 +19,19 @@ type UpdateSchemaRequestMsg struct {
 	Schema []entities.Schema `json:"schema,omitempty"`
 }
 
-// RegisterResponseMsg is sent when receive a register request
+// RegisterResponseMsg represents the outgoing register device response message
```
Comment (netoax) [inline pkg/network/msg.go:22]: I suggest you create another commit to group the comment improvements related to the register operation. You don't need to create one commit per change in each layer, one to group all of them is enough.

**Review by netoax — CHANGES_REQUESTED** [review]

File: pkg/thing/interactors/unregister_thing.go (hunk 7 above)
Comment (netoax) [reply pkg/thing/interactors/unregister_thing.go:33]: Ok, ack.

File: pkg/thing/delivery/http/thing_proxy.go (hunk 16 above)
Comment (esdrasjnr) [reply pkg/thing/delivery/http/thing_proxy.go]: @joaoaneto Done!

File: pkg/network/msg.go (hunk 15 above)
Comment (esdrasjnr) [reply pkg/network/msg.go:5]: @joaoaneto Done!

File: pkg/thing/delivery/http/thing_proxy.go (hunk 17 above)
Comment (esdrasjnr) [reply pkg/thing/delivery/http/thing_proxy.go]: @joaoaneto Done!

File: pkg/network/msg.go (hunk 18 above)
Comment (esdrasjnr) [reply pkg/network/msg.go:22]: @joaoaneto Done!

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
           

Complexity increasing per file
==============================
- pkg/thing/interactors/unregister_thing_test.go  3
- pkg/thing/handler/amqp/msg_handler.go  2
- pkg/thing/interactors/unregister_thing.go  7
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5019174&bid=16546397)

**Review by vitbaq — APPROVED** [review]
LGTM

**Review by vitbaq — APPROVED** [review]

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
           

Complexity increasing per file
==============================
- pkg/thing/interactors/unregister_thing_test.go  3
- pkg/thing/handler/amqp/msg_handler.go  2
- pkg/thing/interactors/unregister_thing.go  7
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5019174&bid=16546397)

File: pkg/thing/interactors/unregister_thing.go (hunk 9 above)
Comment (lcbm) [reply pkg/thing/interactors/unregister_thing.go:27]: Makes sense! :smile: 
Thanks for sharing :wink:

File: pkg/thing/interactors/unregister_thing.go (hunk 10 above)
Comment (lcbm) [reply pkg/thing/interactors/unregister_thing.go:38]: Agreed! :grimacing:

**Review by lcbm — APPROVED** [review]
Looks good to me :tada:

File: pkg/thing/interactors/unregister_thing.go (hunk 9 above)
Comment (netoax) [reply pkg/thing/interactors/unregister_thing.go:27]: I usually prefer to log only when necessary, but this is just my personal preference. Both approaches are correct.
Dave Cheney had written about this subject [here](https://dave.cheney.net/practical-go/presentations/qcon-china.html#_only_handle_an_error_once). I'd go with choosing better where the logs should be placed and avoid redundancy.

**Review by netoax — CHANGES_REQUESTED** [review]
_4fc53e3c244e26837000b26898d7503db8e2d864_
Please, replace the commit title tag from `comment:` to `thing: register:`.
e.g.
`thing: register: Fix comments`
or
`thing: register: Update comments`

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
           

Complexity increasing per file
==============================
- pkg/thing/handler/amqp/msg_handler.go  2
- pkg/thing/interactors/unregister_thing.go  7
- pkg/thing/interactors/unregister_thing_test.go  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5019174&bid=16546397)

**Comment by netoax** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
           

Complexity increasing per file
==============================
- pkg/thing/handler/amqp/msg_handler.go  2
- pkg/thing/interactors/unregister_thing.go  7
- pkg/thing/interactors/unregister_thing_test.go  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5019174&bid=16546397)

**Review by netoax — APPROVED** [review]
